Kernel/Arch/riscv64/SafeMem.cpp

```diff
@@ -12,95 +13,287 @@
+NEVER_INLINE NAKED bool safe_memcpy(void*, void const*, unsigned long, void*&)
 {
-    // FIXME: Actually implement a safe memcpy.
-    auto* pd = static_cast<u8*>(dest_ptr);
-    auto const* ps = static_cast<u8 const*>(src_ptr);
-    for (; n--;)
-        *pd++ = *ps++;
-    fault_at = nullptr;
-    return true;
+    // a0: void* dest_ptr
+    // a1: void const* src_ptr
+    // a2: unsigned long n
+    // a3: void*& fault_at
+
+    asm(R"(
+    beqz a2, 2f
+    add t0, a0, a2         # t0: pointer to the (exclusive) end of the target memory area
+
+1:
+.global safe_memcpy_ins_1
+safe_memcpy_ins_1:
+    lbu t1, (a1)           # t1: byte to copy
+.global safe_memcpy_ins_2
+safe_memcpy_ins_2:
+    sb t1, (a0)
+    addi a0, a0, 1
+    addi a1, a1, 1
+    bne a0, t0, 1b
+
+2:
+    li a0, 1
+.global safe_memcpy_faulted
+safe_memcpy_faulted:
+    ret
+)");
 }
 
 CODE_SECTION(".text.safemem.atomic")
-NEVER_INLINE FLATTEN Optional<bool> safe_atomic_compare_exchange_relaxed(u32 volatile* var, u32& expected, u32 val)
+NEVER_INLINE FLATTEN Optional<bool> safe_atomic_compare_exchange_relaxed(u32 volatile* var, u32& expected, u32 desired)
 {
-    // FIXME: Handle access faults.
-    return AK::atomic_compare_exchange_strong(var, expected, val, AK::memory_order_relaxed);
+    // Based on Example 2, https://github.com/riscv/riscv-isa-manual/releases/download/20240411/unpriv-isa-asciidoc.pdf
+    FlatPtr result;
+    register FlatPtr error asm("t6") = 0;
+    asm volatile(R"(
+    li %[result], 0
+    lw t0, (%[expected_ptr])                       # t0: expected value
+
+1:
+.global safe_atomic_compare_exchange_relaxed_ins_1
+safe_atomic_compare_exchange_relaxed_ins_1:
+    lr.w t1, (%[var_ptr])                          # Load the value at *var into t1.
+    bne t1, t0, 2f                                 # Doesn't match the expected value, so fail.
+.global safe_atomic_compare_exchange_relaxed_ins_2
+safe_atomic_compare_exchange_relaxed_ins_2:
+    sc.w t2, %[desired], (%[var_ptr])              # Try to update the value at *var.
+    bnez t2, 1b                                    # Retry if sc.w failed (that is when t2 != 0).
+    li %[result], 1
+    j 3f
+
+2:
+    sw t1, (%[expected_ptr])                       # Write the read value to expected on failure.
+3:
+.global safe_atomic_compare_exchange_relaxed_faulted
+safe_atomic_compare_exchange_relaxed_faulted:
+)"
+                 : [result] "=&r"(result), "+r"(error)
+                 : [var_ptr] "r"(var), [expected_ptr] "r"(&expected), [desired] "r"(desired)
+                 : "memory", "t0", "t1", "t2");
+    if (error != 0)
+        return {};
+    return static_cast<bool>(result);
 }
 
 CODE_SECTION(".text.safemem.atomic")
 NEVER_INLINE FLATTEN Optional<u32> safe_atomic_load_relaxed(u32 volatile* var)
 {
-    // FIXME: Handle access faults.
-    return AK::atomic_load(var, AK::memory_order_relaxed);
+    u32 result;
+    register FlatPtr error asm("t6") = 0;
```

Comment:
I think it would be pertinent to inform readers that this "temporary" register name is not chosen randomly and must agree between all safe_atomic functions and the safe fault handler.